### PR TITLE
Run scheduled Task only in Processing Phase

### LINKED ISSUE #8642: NPE when scheduling Actor to check pending process message subscriptions
**Describe the bug**

A NullPointerException occurred when scheduling an actor to check for pending process message subscriptions in a 30 second interval.
```java
private void scheduleTimer() {
    if (timer == null) {
      timer = actor.runAtFixedRate(SUBSCRIPTION_CHECK_INTERVAL, this::checkPendingSubscriptions);
    }
  }
```
In this case actor was null. This seems to have occurred when the broker was being shutdown.

**To Reproduce**

Not tried. It seems to have happened by shutting down the cluster.

**Expected behavior**

**Log/Stacktrace**

<details><summary>Full Stacktrace</summary>
 <p>

```
java.lang.NullPointerException: Cannot invoke "io.camunda.zeebe.util.sched.ActorControl.runAtFixedRate(java.time.Duration, java.lang.Runnable)" because "this.actor" is null
	at io.camunda.zeebe.engine.processing.message.PendingProcessMessageSubscriptionChecker.scheduleTimer(PendingProcessMessageSubscriptionChecker.java:69) ~[zeebe-workflow-engine-1.3.0.jar:1.3.0]
	at io.camunda.zeebe.engine.processing.message.PendingProcessMessageSubscriptionChecker.onResumed(PendingProcessMessageSubscriptionChecker.java:64) ~[zeebe-workflow-engine-1.3.0.jar:1.3.0]
	at java.util.ArrayList.forEach(Unknown Source) ~[?:?]
	at io.camunda.zeebe.engine.processing.streamprocessor.StreamProcessor.lambda$resumeProcessing$11(StreamProcessor.java:444) ~[zeebe-workflow-engine-1.3.0.jar:1.3.0]
	at io.camunda.zeebe.util.sched.ActorControl.lambda$call$0(ActorControl.java:136) ~[zeebe-util-1.3.0.jar:1.3.0]
	at io.camunda.zeebe.util.sched.ActorJob.invoke(ActorJob.java:62) ~[zeebe-util-1.3.0.jar:1.3.0]
	at io.camunda.zeebe.util.sched.ActorJob.execute(ActorJob.java:39) [zeebe-util-1.3.0.jar:1.3.0]
	at io.camunda.zeebe.util.sched.ActorTask.execute(ActorTask.java:122) [zeebe-util-1.3.0.jar:1.3.0]
	at io.camunda.zeebe.util.sched.ActorThread.executeCurrentTask(ActorThread.java:95) [zeebe-util-1.3.0.jar:1.3.0]
	at io.camunda.zeebe.util.sched.ActorThread.doWork(ActorThread.java:78) [zeebe-util-1.3.0.jar:1.3.0]
	at io.camunda.zeebe.util.sched.ActorThread.run(ActorThread.java:192) [zeebe-util-1.3.0.jar:1.3.0] 
```

</p>
</details>

<details><summary>Surrounding logs</summary>
 <p>
https://console.cloud.google.com/logs/query;query=logName:%22stdout%22%0Aresource.type%3D%22k8s_container%22%0Aresource.labels.pod_name%3D%22zeebe-1%22%0Aresource.labels.namespace_name%3D%222d7cd1f4-2711-4f9c-8cf7-7b8d887970df-zeebe%22%0Aresource.labels.cluster_name%3D%22production-worker-1%22%0Aresource.labels.project_id%3D%22camunda-cloud-240911%22%0Aresource.labels.container_name%3D%22zeebe%22%0Aresource.labels.location%3D%22europe-west1%22%0Alabels.%22k8s-pod%2Fstatefulset_kubernetes_io%2Fpod-name%22%3D%22zeebe-1%22%0Alabels.%22k8s-pod%2Fapp_kubernetes_io%2Fcomponent%22%3D%22gateway%22;timeRange=2022-01-24T18:30:34.458Z%2F2022-01-24T19:30:34.458Z;pinnedLogId=2022-01-24T19:00:04.458609818Z%2Fzsv4lnn1az3kwuxb;cursorTimestamp=2022-01-24T19:00:06.140812079Z?project=camunda-cloud-240911

```
Info 2022-01-24 20:00:04.243 CET zeebe"Prepare transition from FOLLOWER on term 26 to INACTIVE - preparing LogStorage"
Info 2022-01-24 20:00:04.243 CET zeebe"Preparing transition from FOLLOWER on term 26 completed"
Info 2022-01-24 20:00:04.243 CET zeebe"Transition to INACTIVE on term 26 starting"
Info 2022-01-24 20:00:04.243 CET zeebe"Transition to INACTIVE on term 26 - transitioning LogStorage"
Info 2022-01-24 20:00:04.243 CET zeebe"Transition to INACTIVE on term 26 - transitioning LogStream"
Info 2022-01-24 20:00:04.243 CET zeebe"Transition to INACTIVE on term 26 - transitioning ZeebeDb"
Info 2022-01-24 20:00:04.244 CET zeebe"Transition to INACTIVE on term 26 - transitioning QueryService"
Info 2022-01-24 20:00:04.244 CET zeebe"Transition to INACTIVE on term 26 - transitioning StreamProcessor"
Info 2022-01-24 20:00:04.244 CET zeebe"Transition to INACTIVE on term 26 - transitioning SnapshotDirector"
Info 2022-01-24 20:00:04.244 CET zeebe"Transition to INACTIVE on term 26 - transitioning ExporterDirector"
Info 2022-01-24 20:00:04.244 CET zeebe"Transition to INACTIVE on term 26 completed"
Debug 2022-01-24 20:00:04.245 CET zeebe"Shutdown was called with context: io.camunda.zeebe.broker.system.partitions.PartitionStartupAndTransitionContextImpl@6b186dc6"
Info 2022-01-24 20:00:04.245 CET zeebe"Shutdown RocksDB metric timer"
Info 2022-01-24 20:00:04.245 CET zeebe"Shutdown LogDeletionService"
Debug 2022-01-24 20:00:04.246 CET zeebe"Finished shutdown process"
Debug 2022-01-24 20:00:04.357 CET zeebe"Closed database from '/usr/local/zeebe/data/raft-partition/partitions/2/runtime'."
Debug 2022-01-24 20:00:04.426 CET zeebe"Received metadata change from Broker 1, partitions {1=FOLLOWER, 2=FOLLOWER}, terms {} and health {1=UNHEALTHY, 2=UNHEALTHY}."
Debug 2022-01-24 20:00:04.440 CET zeebe"Configured log appender back pressure at partition 1 as AppenderVegasCfg{initialLimit=1024, maxConcurrency=32768, alphaLimit=0.7, betaLimit=0.95}. Window limiting is disabled"
Debug 2022-01-24 20:00:04.441 CET zeebe"Recovering state of partition 1 from snapshot"
Info 2022-01-24 20:00:04.442 CET zeebe"Engine created. [value-mapper: CompositeValueMapper(List(io.camunda.zeebe.el.impl.feel.MessagePackValueMapper@72f68c0)), function-provider: io.camunda.zeebe.el.impl.feel.FeelFunctionProvider@7d7a19fa, clock: io.camunda.zeebe.el.impl.ZeebeFeelEngineClock@57013eea, configuration: Configuration(false)]"
Info 2022-01-24 20:00:04.444 CET zeebe"Recovered state of partition 1 from snapshot at position -1"
Info 2022-01-24 20:00:04.444 CET zeebe"Engine created. [value-mapper: CompositeValueMapper(List(io.camunda.zeebe.el.impl.feel.MessagePackValueMapper@1c5b8eb)), function-provider: io.camunda.zeebe.el.impl.feel.FeelFunctionProvider@7dd0b97b, clock: io.camunda.zeebe.el.impl.ZeebeFeelEngineClock@5048caf2, configuration: Configuration(false)]"
Info 2022-01-24 20:00:04.445 CET zeebe"Engine created. [value-mapper: CompositeValueMapper(List(io.camunda.zeebe.el.impl.feel.MessagePackValueMapper@ee0587f)), function-provider: io.camunda.zeebe.el.impl.feel.FeelFunctionProvider@22cebba8, clock: io.camunda.zeebe.el.impl.ZeebeFeelEngineClock@5a2dc5eb, configuration: Configuration(false)]"
Info 2022-01-24 20:00:04.446 CET zeebe"Engine created. [value-mapper: CompositeValueMapper(List(io.camunda.zeebe.el.impl.feel.MessagePackValueMapper@4c249b06)), function-provider: io.camunda.zeebe.el.impl.feel.FeelFunctionProvider@400b015d, clock: io.camunda.zeebe.el.impl.ZeebeFeelEngineClock@174ff10f, configuration: Configuration(false)]"
Info 2022-01-24 20:00:04.448 CET zeebe"Processor starts replay of events. [snapshot-position: -1, replay-mode: PROCESSING]"
Debug 2022-01-24 20:00:04.448 CET zeebe"Detected 'HEALTHY' components. The current health status of components: [raft-partition-partition-1{status=HEALTHY}, Broker-1-LogStream-1{status=HEALTHY}, Broker-1-StreamProcessor-1{status=HEALTHY}]"
Info 2022-01-24 20:00:04.448 CET zeebe"Processor finished replay, with [lastProcessedPosition: -1, lastWrittenPosition: -1]"
Info 2022-01-24 20:00:04.448 CET zeebe"Cancelling transition to LEADER on term 28"
Debug 2022-01-24 20:00:04.451 CET zeebe"Paused processing for partition 1"
Error 2022-01-24 20:00:04.458 CET zeebe"Actor Broker-1-StreamProcessor-1 failed in phase STARTED."
```

</p>
</details>

**Environment:**
 - OS: Camunda Cloud
 - Zeebe Version: 1.3.0
 - Cluster: 2d7cd1f4-2711-4f9c-8cf7-7b8d887970df (cluster is not available anymore)

### LINKED ISSUE #9962: Pausing a partition should pause all processing, not just the processing state machine
**Describe the bug**

When pausing processing of a partition which is already processing, it pauses only the processing state machine. Anything outside of it but running in the same actor may still be working, e.g. the distribution behavior is still redistributing. 

**To Reproduce**

Start a cluster with a least two partition, and make sure you have different leaders for each. Then pause processing on the second partition, and deploy a BPMN on the first. You will see that it keeps retrying since there is no answer.

Now pause processing of partition 1, and you will see it still keeps retrying.

**Expected behavior**

Pausing processing pauses the complete engine, not just the processing state machine. That includes redistributing deployments, timer triggers, message subscriptions, etc. We cannot pause the complete actor obviously since we may need to resume, but no work should be done in general.

### LINKED ISSUE #9999: ProcessingScheduleService allows to schedule a task  which returns a Result
**Description**

We started to introduce the new `ProcessingScheduleService` still the consumers of this service use the LogStreamWriter, which they get from the Engine/StreamProcessorContext. 

Ideally the consumers of the Service can just schedule a Task, which returns an Result. The result should be written to the Log (internally in the streamProcessor). The consumer of the service doesn't care how this is done, and doesn't need to know about LogStreamWriters. 

What we need:

- [x] Introduce a ProcessingScheduleResult
  - [x] potentially backed by similar implementation (parts) of the other ProcessingResult
- [x] New methods to schedule a Task which returns an result
- [x] Implementation of that
- [x] Backened in the StreamProcessor, like writing the result to the logstream
- [x] We need to make sure that the scheduling is guaranteed to happen only during processing not in between and after a committed transaction
- [ ] During pause no execution should happen
- [ ] We need tests

related to https://github.com/camunda/zeebe/issues/9730

### PR DESCRIPTION
**Pull request by ChrisKujawa**
## Description
This PR makes sure that we no longer run Tasks in other phases than in Processing. 

>     We want to execute the scheduled tasks only if the StreamProcessor is in the PROCESSING
>     PHASE, but no processing is currently happening.
>     
>     To make sure that:
>     
>      * we are not running during replay/init phase (the state might not be up-to-date yet)
>      * we are not running during suspending
>      * we are not interfering with the current ongoing processing,
>        such that all transaction changes are available during our task execution
> 

This means when the StreamProcessor is suspended scheduled Tasks are not executed they are put back to the queue this was the simplest solution for now, since we have no Actor suspend feature for now #10006. Furthermore, we no longer run scheduled tasks on replay since this might also lead to issues (because the state is not yet fully rebuild etc.)

Added tests for certain aspects of this feature.
    

## Related issues

closes https://github.com/camunda/zeebe/issues/9962
closes https://github.com/camunda/zeebe/issues/9999
closes https://github.com/camunda/zeebe/issues/8642 (this will no longer happen since the abort condition is also verified in the SchedulingService)

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [x] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/camunda/zeebe/compare/stable/0.24...main?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/1.3`) to the PR, in case that fails you need to create backports manually.

Testing:
* [x] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark

Documentation:
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] New content is added to the [release announcement](https://drive.google.com/drive/u/0/folders/1DTIeswnEEq-NggJ25rm2BsDjcCQpDape)
* [ ] If the PR changes how BPMN processes are validated (e.g. support new BPMN element) then the Camunda modeling team should be informed to adjust the BPMN linting.

Please refer to our [review guidelines](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews#code-review-guidelines).

### CODE REVIEW

**Review by npepinpe — COMMENTED** [review]
Looks good in general.

I have one question about the requirement:

> we are not interfering with the current ongoing processing, such that all transaction changes are available during our task execution

I thought if everything worked within its own transaction then it shouldn't interfere with each other anyway, no? Maybe I misunderstood how transactions work in RocksDB. Can you explain that to me?

Anyway, so right now it means if I want to schedule something, I will schedule it anyway but might not do anything, and we rely on the `runAtFixedRate` implementation to reschedule it, correct? So in a paused service, this means it will keep enqueuing forever until it's unpaused, correct? Could that cause starvation to the things that run at fixed rate?

e.g. I want to do something at fixed rate, but every time I try to schedule, we're currently in processing, so I never do anything. 

How likely is this? Unlikely? Cause in a processing system, I expect we're very often "in processing", right?

File: engine/src/main/java/io/camunda/zeebe/streamprocessor/StreamProcessorContext.java [hunk 1]
```diff
@@ -61,6 +62,9 @@ public void onSkipped(final LoggedEvent skippedRecord) {}
   // safe structure here
   private boolean inProcessing;
 
+  // this is accessed outside, which is why we need to make sure that it is thread-safe
+  private volatile StreamProcessor.Phase phase = Phase.INITIAL;
```
[the remark this answers is not in the document]
Comment (Zelldon) [reply engine/src/main/java/io/camunda/zeebe/streamprocessor/StreamProcessorContext.java:66]: Right now the health monitir and I don't want to fix this in this PR so I kept as it is

**Comment by Zelldon** [comment]
The stream processor doesn't use any runAtFixed rate anymore. This is just a wrapper around runDelayed and rescheduling.

Regarding never executed this should also no issue since we do not use #run in the stream processor anymore, which means everything will appended to the same queue and no skip the line is happening. 

Hope that helps.

**Comment by npepinpe** [comment]
> Regarding never executed this should also no issue since we do not use #run in the stream processor anymore, which means everything will appended to the same queue and no skip the line is happening.

Maybe I misunderstood, but even if `runAtFixedRate` is just a wrapper around `runDelayed`, there's still the possible case of:

1. Process is inProcessing
2. Schedule runnable X
3. Processor update states (still inProcessing)
4. Scheduled runnable X does nothing because inProcessing; gets rescheduled for the next delay
5. Processor does more stuff, etc.
6. Processor gets back to inProcessing
7. Delay passes, so now the runnable X executes; does nothing because inProcessing, gets rescheduled for next delay

Couldn't this happen in a loop? My question was, do we consider this so unlikely (that it would loop like this)? Cause I would expect we're most of the time "in processing", right? We're only not in processing between executeSideEffects and processCommand (so only when reading the next record). Since scheduled jobs aren't sequenced with normal jobs (they get sequenced only when the timer expires), there's no guarantee of ordering? Am I missing something?

**Comment by npepinpe** [comment]
> P.S. regarding the transaction the requirement came up because the engine wants to schedule jobs (replacement for sideeffects) but they might be not committed yet.

I really thought RocksDB transactions were isolated from each other. Is that not true? Or is it because side effects (or the replacement) use the same transaction context as the processing?

**Comment by Zelldon** [comment]
> I really thought RocksDB transactions were isolated from each other. Is that not true? Or is it because side effects (or the replacement) use the same transaction context as the processing?

I think we talk about two things here. It is not about whether transactions are isolated or not. It is about data should be written to the state before the state is access again on a scheduled task. For example and to be specific https://github.com/camunda/zeebe/blob/main/engine/src/main/java/io/camunda/zeebe/engine/processing/common/CatchEventBehavior.java#L302-L308 we want that this to be executed AFTER the transaction is committed. Right now this is done via side-effects, which we wanted to replace with the SchjedulingService as you can see here https://github.com/camunda/zeebe/issues/9723#issuecomment-1177371393


But since we talk about it: @pihme created during I was on vacation the post-commit tasks https://github.com/camunda/zeebe/blob/main/engine/src/main/java/io/camunda/zeebe/engine/api/PostCommitTask.java which can be added to the ProcessingResult. We could remove that with the ProcessingScheduleService after this PR, so we would need to maintain only one approach. OR we decide that we keep them and use it for that use case and remove from the ProcessingScheduleService the property that we guarantee the case that it is not executed during inProcessing. Worth to discuss I guess.
@pihme

**Comment by Zelldon** [comment]
> > Regarding never executed this should also no issue since we do not use #run in the stream processor anymore, which means everything will appended to the same queue and no skip the line is happening.
> 
> Maybe I misunderstood, but even if `runAtFixedRate` is just a wrapper around `runDelayed`, there's still the possible case of:
> 
> 1. Process is inProcessing
> 2. Schedule runnable X
> 3. Processor update states (still inProcessing)
> 4. Scheduled runnable X does nothing because inProcessing; gets rescheduled for the next delay
> 5. Processor does more stuff, etc.
> 6. Processor gets back to inProcessing
> 7. Delay passes, so now the runnable X executes; does nothing because inProcessing, gets rescheduled for next delay
> 
> Couldn't this happen in a loop? My question was, do we consider this so unlikely (that it would loop like this)? Cause I would expect we're most of the time "in processing", right? We're only not in processing between executeSideEffects and processCommand (so only when reading the next record). Since scheduled jobs aren't sequenced with normal jobs (they get sequenced only when the timer expires), there's no guarantee of ordering? Am I missing something?

Not really. The runnable is scheduled with the delay and after the delay is done the inProcessing is checked. IF inProcessing is true we add it to the queue, and eventually we will execute it since all jobs are added to the same queue (no delay is here again).

**Comment by Zelldon** [comment]
> OR we decide that we keep them and use it for that use case and remove from the ProcessingScheduleService the property that we guarantee the case that it is not executed during inProcessing. 

I guess this would be also interesting, since we have not the issue like here https://github.com/camunda/zeebe/pull/10112#issuecomment-1220967877 that runDelayed and runAtFixedRate are postPoned for too long, because the StateMachine is in processing.

**Comment by npepinpe** [comment]
My bad, I must've been tired when I did the review because it seems pretty obvious now. So we schedule, and if we're "busy", then we enqueue the task for later consumption. Yeah it would be nice if we don't re-schedule immediately but instead only when the task has actually run, cause right now it might cause two tasks which should be scheduled with a delay in between to be run immediately one after the other :shrug: 

Regarding the transaction, if I understand correctly, it's just about making sure that the scheduled task isn't triggered before the current record processing is finished (since it needs access to whatever will be committed). Sorry, I'm a little slow right now :sweat_smile: 

I'm not familiar with the decisions to remove the side effects/post commit tasks, to be honest, so I'll trust that you all took the best decision and that it's the right thing to do. 

I think it's not hard to modify the `runAtFixedRate` implementation to make sure we only re-schedule when the task is actually executed, e.g.

```java
  @Override
  public void runAtFixedRate(final Duration delay, final Task task) {
    /* TODO preliminary implementation; with the direct access
     * this only works because this class is scheduled on the same actor as the
     * stream processor.
     */
    runDelayed(
        delay,
        toRunnable(
            builder -> {
              try {
                return task.execute(builder);
              } finally {
                runAtFixedRate(delay, task);
              }
            }));
  }
```

This can probably be "optimized" to avoid rebuilding the runnable/rescheduling task I guess, but you get the idea.

**Review by npepinpe — APPROVED** [review]
:+1: 

I'm a little tired and slower these days, sorry for the back and forth, it makes sense to me now.

I think we can improve by only rescheduling the task after it's been actually executed (and not immediately after the converted `Runnable` runs). That's simple enough to do I think, so no need for another review.

**Comment by Zelldon** [comment]
thanks @npepinpe I did that

bors r+

**Pull request merged**